--- link.ld	2026-04-14 17:06:31.145859809 +0300
+++ link_patch4.ld	2026-04-14 17:06:24.606021579 +0300
@@ -82,3 +82,36 @@
         _stack_bottom = _stack_start - _min_stack;
     } > ram
 }
+
+/* ==================================================================== */
+/* PATCH_04 BEGIN                                                       */
+/*                                                                      */
+/* REASON:  need runtime variables from the C code to know the exact    */
+/*          size of imem and dmem to not accidentelly write to valid    */
+/*          memory values using cli.                                    */
+/*                                                                      */
+/* AUTHOR:  Manfredas Lamsargis                                         */
+/*                                                                      */
+/* DATE:    14.04.26                                                    */
+/* ==================================================================== */
+
+/* --- Program Instruction Segment (IMEM) --- */
+__prog_instr_start = ADDR(.vectors);
+__prog_instr_end   = ADDR(.rodata) + SIZEOF(.rodata);
+__prog_instr_size  = __prog_instr_end - __prog_instr_start;
+
+/* --- Program Data Segment (DMEM) --- */
+__prog_data_start    = ADDR(.data);
+__prog_data_end      = ADDR(.stack) + SIZEOF(.stack);
+__prog_data_ram_size = __prog_data_end - __prog_data_start;
+__prog_data_bin_size = SIZEOF(.data);
+
+/* --- Memory Segments --- */
+__size_text     = SIZEOF(.text);
+__size_data     = SIZEOF(.data);
+__size_bss      = SIZEOF(.bss);
+__size_stack    = SIZEOF(.stack);
+
+/* ==================================================================== */
+/* PATCH_04 END                                                         */
+/* ==================================================================== */
