--- common.mk	2026-02-23 13:25:42.306204741 +0200
+++ common-new.mk	2026-02-23 13:25:33.685104924 +0200
@@ -30,10 +30,10 @@
 #########################
 
 # Architecture-specific options for the assembler.
-AOPS = -march=rv32imc_zicsr
+AOPS = -march=rv32imc
 
 # Compilation options for C files.
-COPS = $(AOPS) -mabi=ilp32 -Wall -O2 -Os -Oz -nostdlib -nostartfiles -ffreestanding -fdata-sections -g 
+COPS = $(AOPS) -mabi=ilp32 -Wall -nostdlib -nostartfiles -ffreestanding -fdata-sections -g 
 
 
 # Add the common include directory to the compiler flags.
